### DISCUSSION TITLE
Set random seed?

### DISCUSSION THREAD

**Discussion by fohria** [jspsych/jsPsych#626] [Q&A]
I'm currently using the 'fixed-repetitions' sample function, to randomize three kinds of trials an equal amount of times during 30 total trials.

I'd like the created sequence to be the same for all participants in my experiment, so I thought I could just set a random seed, but apparently that doesn't exist natively in javascript (https://stackoverflow.com/questions/521295/seeding-the-random-number-generator-in-javascript)

It looks like there's a semi standard in the js community to use this library for enabling setting a seed; https://github.com/davidbau/seedrandom

So I could use that with a custom sample function to enable what I need.

Is there some other way to achieve this with the current functionality in jsPsych that I've overlooked?

**Comment by jodeleeuw** [comment]
I think your strategy is probably the correct one. Since you want a fixed sequence, you could also generate the sequence once using the built in RNG locally, and hard code it into a custom sample function.

**Comment by fohria** [comment]
Using seedrandom was much more straight forward than I thought. It can overwrite the Math.random() function globally, if you set it like this on the first line of your jsPsych script:

`Math.seedrandom('whateverstringyouwant');`

The timeline then automagically becomes perfectly predictable.

The seedrandom readme does state that this is insecure for any cryptography or similar use cases, but for my use case I don't think it really matters.

The recommended way is to create a separate random generator:

`var myrng = new Math.seedrandom('hello.');`

I think it would be fairly straightforward to allow a setting in jsPsych for setting a seed in the sample functions; if set to a string, it uses `myrng` with that string instead of Math.Random in the randomization function(s). If it's not set or maybe set to false, it uses the current code with Math.Random.

Then again maybe it would be awkward to require downloading a separate library to use the seed function, what do you think?

**Comment by jodeleeuw** [comment]
Glad to hear that this works well!

I've been hesitant to bundle in a separate library like this, especially for a function that would be used only occasionally. 

I suppose that the library is small enough that it could just be merged into jsPsych. It's MIT Licensed. On the other hand, given how easy it is to do this as you demonstrated, it may not be worth it.

**Comment by fohria** [comment]
Yeah I see your point.

I'll add a little extra info here then for anyone else searching for setting a seed:

1. Download seedrandom.min.js from https://github.com/davidbau/seedrandom
2. Add it to your project folder
3. Add it to your experiment file, for example:
```
<title>My experiment</title>
        <script src="seedrandom.min.js"></script>
        <script src="jspsych-6.0.5/jspsych.js"></script>
        <script src="jspsych-6.0.5/plugins/jspsych-html-keyboard-response.js"></script>
        <link href="jspsych-6.0.5/css/jspsych.css" rel="stylesheet" type="text/css"></link>
</title>
```
4. On the first line of your experiment script section, add 'Math.seedrandom('whateverstringyouwant');' like so:
```
    <script>

    Math.seedrandom('whateverstringyouwant');

    var hello_trial = {
        type: 'html-keyboard-response',
        stimulus: 'Hello world!'
    }

    // more experiment javascript code

    </script>
```

Now any sample function you use (https://www.jspsych.org/overview/timeline/#sampling-methods) will be the same every time!

**Comment by j22tis** [comment]
Hi! 

Thanks for this approach! 

I tried to use it (added Math.seedrandom('river'); right after  <script>) in my experiment where there are two kinds of blocks and the block size is randomly sampled according to a this randomization scheme: 
var easy_block = {
     timeline: [experimental_trial],
     timeline_variables: easy_params,
     sample: {
type: "custom",
fn: function(t) {
var block_size = jsPsych.randomization.sampleWithReplacement([15,16,17,18,19,20],1)[0];
var t = jsPsych.randomization.sampleWithReplacement(t, block_size, [0.25,0.25,0.75,0.75]);
return t}
}

// Construct a hard block //
var hard_block = {
         timeline: [experimental_trial],
         timeline_variables: hard_params,
         sample: {
    type: "custom",
    fn: function(t) {
  var block_size = jsPsych.randomization.sampleWithReplacement([15,16,17,18,19,20],1)[0];
  var t = jsPsych.randomization.sampleWithReplacement(t, block_size, [0.25,0.25,0.75,0.75]);
  return t
    }
  },

The goal is to create the same sequence of trials for every participant. 

When I plot the outcome of a pilot with 5 participants, it seems that the block length is the same for the first two blocks, but then it changes for some reason: 
![image](https://user-images.githubusercontent.com/22985804/119471650-c4fb5500-bd49-11eb-986b-6791a0056e95.png)

Any idea, why this might be happening? Are there some newer approaches for creating a randomized trial sequence beforehand so it would be the same for each participant?
